### DISCUSSION TITLE
Interesting behaviour with duplicate load options

### DISCUSSION THREAD

**Discussion by michaelbukachi** [sqlalchemy/sqlalchemy#8164] [Usage Questions]
So I discovered some very weird behaviour that I don't know how to explain. Let's you have the following snippet:
```
class User:
    __tablename__ = 'users'
    id = Column(Integer)
    name = Column(String)
    phone = Column(String)

class Task:
    __tablename__ = 'tasks'
    id = Column(Integer)
    created_by_id = Column(Integer, ForeignKey("users.id"))
    managed_by_id = Column(Integer, ForeignKey("users.id"))
    created_by: "User" = relationship(
        "User", foreign_keys=[created_by_id]
    )
   managed_by: "User" = relationship(
        "User", foreign_keys=[managed_by_id]
    )
```
And we try to loading tasks using:
```
session.query(Task).options(
    joinedload(Task.created_by).load_only(User.name, User.phone),
    joinedload(Task.manged_by).load_only(User.name)
).all()
```
If the `managed_by` is accessed before `created_by`, and the later we try to access `task.created_by.phone`. It triggers a lazy load (which can be fatal considering async). This only happens for scenarios where `created_by` and `managed_by` are the same user.

Can anyone explain this? :thinking:  Are there strategies to avoid this like maybe always loading the obj with more options (I know, very hacky :sweat_smile: ).

**Comment by zzzeek** [comment]
I think it's amazing it works correctly when the User objects are different (as it should of course), if I think about how complicated that is.

When they're the same object, but loaded from different paths (that is, one from created_by, one from managed_by), the loading process is as you might imagine pretty intricate and I'm not really sure where to go with that, at the moment I'm having trouble reproducing that situation consistently (but I did reproduce something like what you are getting, probably, a few times, it just isn't doing it now).

When it loads things, it iterates through each row, pulls out a Task from it, then pulls out "managed_by" and "created_by", and the order in which it does that is not really deterministic, then loads that object and its members based on what it was instructed to load, either "name/phone" or "name".

so at some point, the loader to populate "phone" might be first, then a "deferred" loader comes and cancels it out, or the "deferred" loader populates it, then the loader to populate "phone" would skip the object because it's already populated.

I guess there could be a rule to try to work this out correctly as we get that object.  However if you are able to reproduce this consistently try seeing if adding populate_existing() to your query improves things.

**Reply by michaelbukachi** [reply]
I'm unable to produce deterministic results as well. It just happens randomly. A quick workaround is to make sure the same load options are passed for a model in a query. It's definitely not a solution but at least asyncio doesn't crash as often due to implicit io.

**Reply by zzzeek** [reply]
well OK there is a place where we get the object and we should be able to tell everything about it.   I have a case that can sometimes reproduce it and there's a definite place logic can be taken to see, "we have a deferred loader for 'phone', but this loader here has an actual value, let's use it".  ill make an issue for that.

populate_existing() right now might solve the problem though.

**Reply by zzzeek** [reply]
I have a patch that I'd rather keep for 2.0, it's from #8166.

**Reply by michaelbukachi** [reply]
Let me test with `populate_existing`, if it's indeed a fix then an exception won't be raised no matter how many times I rerun the test tes.
